---
title: "runDRIM+DEXseq"
output: 
  html_document:
    df_print: paged
---

N.b. this script must be run in the "R-rstudio" environment:
"/shared/sudlab1/General/apps/conda/conda-install/envs/R-rstudio"

Import necessary libraries

```{r import_libs, warning=FALSE }
library(dplyr, quietly=T)
library(tximport, quietly=T)
library(DESeq2, quietly=T)
library(stringr, quietly=T)
library(ggplot2)
library(GenomicFeatures)
library(DRIMSeq)
library(stageR)
library(DEXSeq)
library(tidyr)
library(data.table)
```

Load in the configs / run-specific information

```{r load_configs}
#load config and info file
pipeline_yml = yaml::read_yaml("../../pipeline.yml")

#determine which comparison we are currently doing
folder_name = basename(getwd())

#get formula
formula = read.table(paste0(folder_name, ".txt")) %>% unlist() %>% unname()

#we need to know what var1 values to filter by
filter_by_var1 = str_split(folder_name, "_in_")[[1]][1] %>% str_split("_vs_") %>% unlist()

#if there are 2 'vs' in the name then we are doing the
#interaction comparison, i.e. ~X + Y + X:Y
do_interaction = ifelse((length(grep("vs", str_split(folder_name, "_")[[1]])) == 2), TRUE, FALSE)

#if there is an 'in' present then we will need to filter
#the sample_table later, so we need to know if so
do_filter_by_var2 = "in" %in% str_split(folder_name, "_")[[1]]

#we also need to know what to filter by, e.g. wt or mutant, or both
if (do_filter_by_var2) {
  filter_by_var2 = str_split(folder_name, "_in_")[[1]][2] %>% str_split("_vs_") %>% unlist()
}
```

Generate sample table for tximport

```{r sample_table_from_sf}
#list all salmon files in quantification directory

quant_dir = "../../quantification.dir"

file_namings = fread("../../file_naming.tsv", header=F)

comparison = as.character(str_split(basename(getwd()), "_vs_")[[1]])
file_namings = file_namings %>% filter(V1 %in% comparison)
condition1_files = list.files("../../quantification.dir", pattern=as.character(file_namings[1,2]), recursive=T)
condition2_files = list.files("../../quantification.dir", pattern=as.character(file_namings[2,2]), recursive=T)

condition1_files = condition1_files[grepl("/", condition1_files)]
condition2_files = condition2_files[grepl("/", condition2_files)]

condition1_filenames = sapply(condition1_files, function(x) str_split(substr(x, 0, nchar(x)-3),"/")[[1]][1])
condition2_filenames = sapply(condition2_files, function(x) str_split(substr(x, 0, nchar(x)-3),"/")[[1]][1])

condition1_files = paste0(quant_dir, "/", condition1_files)
condition2_files = paste0(quant_dir, "/", condition2_files)

#create basic sample table for condition 1 with var1 set to comparison[1].
sample_table_1 = data.frame(files=condition1_files, sample_id=condition1_filenames) %>% 
  mutate(files = as.character(condition1_files)) %>% mutate(var1 = comparison[1]) %>%
  transmute(sample_id = sample_id,
            var1 = var1,
            files = files)

#create basic sample table for condition 2 with var1 set to comparison[2].
sample_table_2 = data.frame(files=condition2_files, sample_id=condition2_filenames) %>% 
  mutate(files = as.character(condition2_files)) %>% mutate(var1 = comparison[2]) %>%
  transmute(sample_id = sample_id,
            var1 = var1,
            files = files) 

#merge the 2 sample tables
sample_table = rbind(sample_table_1, sample_table_2)
```

How many samples are in each condition?

```{r overview_table}
table(sample_table$var1)
```

Check the files exist

```{r check_files_exist}
all(file.exists(sample_table$files))
```
 
Created named list of files

```{r named_list_of_files}
files = sample_table$files
names(files) = sample_table$sample_id
head(files)
```

Import with tximport

```{r tximport}
txi = tximport(files, type="salmon", txOut=TRUE, countsFromAbundance="no")
cts = txi$counts
#remove any rows where there are no counts
cts = cts[rowSums(cts) > 0,]
```

### Transcript to gene mapping

Load tx2gene from csvdb_files and rename so it matches quantification convention

```{r load_txdf}
txdf = read.delim("../../expression.dir/csvdb_files/tx2gene.txt")
txdf = txdf %>% transmute(GENEID = match_gene_id, TXNAME = transcript_id)
tab = table(txdf$GENEID)
txdf$ntx = tab[match(txdf$GENEID, names(tab))]
```

We need nrows to be same for both the txdf and cts so we need to manipulate both (just in case they differ, which they likely will now due to having filtered out transcripts with count == 0)

```{r reduce_txdf}
nrow(txdf)
nrow(cts)

cts.names=row.names(cts)
in_cts_not_txdf = base::setdiff(cts.names, txdf$TXNAME)
in_txdf_not_cts = base::setdiff(txdf$TXNAME, cts.names)

cts = cts[!(row.names(cts) %in% in_cts_not_txdf),]
txdf = txdf %>% filter(!TXNAME %in% in_txdf_not_cts)

nrow(txdf)
nrow(cts)
```

Now they should be the same!

### DRIMSeq

Create counts matrix with each row as a transcript, and each column as a sample

```{r count_matrix}
library(data.table)
counts = data.frame(cts)
counts = setDT(counts, keep.rownames = TRUE)
colnames(counts)[which(names(counts) == "rn")] = "feature_id"
counts = merge(counts, txdf, by.x = "feature_id", by.y = "TXNAME")
counts = counts[,1:(length(counts)-2)]
counts = data.frame(counts)
counts = counts %>% relocate(GENEID) 
colnames(counts)[which(names(counts) == "GENEID")] = "gene_id"
row.names(counts) = counts$feature_id
```

For some reason this has changed colnames from identifier-var1-replicate to identifier.var1.replicate. So lets change both to replace all dots with dashes.

```{r rename_count_matrix_heads}
names(counts) = stringr::str_replace_all(names(counts), "\\.", "\\-")
sample_table = sample_table %>% mutate(sample_id = stringr::str_replace_all(sample_id, "\\.", "\\-"))
```

Additionally, there appear to be a few MSTRG transcripts that have no associated gene... so filter those away

```{r filter_novel_no_genes}
counts = counts %>% filter(gene_id != "")
```

Create dmDSdata object

```{r create_dmDSdataObject}
d = dmDSdata(counts=counts, samples=sample_table)
d
```

```{r head_dmDSdataObject}
counts(d[1,])
```

DRIM/DEX filtering based on pipeline.yml

```{r DTU_filtering}
n = dim(counts)[2]-2
tx_n = round(n * as.numeric(pipeline_yml$dtu$drim_dex_filters$minimum_samples_transcript_n))
tx_expr = as.numeric(pipeline_yml$dtu$drim_dex_filters$minimum_samples_tx_expression_cutoff)
tx_prop = as.numeric(pipeline_yml$dtu$drim_dex_filters$minimum_samples_tx_proportion_cutoff)
gene_n = round(n* as.numeric(pipeline_yml$dtu$drim_dex_filters$minimum_samples_gene_n))
gene_expr = as.numeric(pipeline_yml$dtu$drim_dex_filters$minimum_samples_gene_expression)

d = dmFilter(d,
             min_samps_feature_expr=tx_n, min_feature_expr=tx_expr,
             min_samps_feature_prop=tx_n, min_feature_prop=tx_prop,
             min_samps_gene_expr=(gene_n), min_gene_expr=gene_expr)
d
```

How many isoforms are there for each gene?

```{r table_d}
table(table(counts(d)$gene_id))
```

Now to create the design matrix based on design_tsv complexity 

```{r design_formula}
if(formula == "~var1"){
  design_full = model.matrix(~var1, data=DRIMSeq::samples(d))
}else if(formula == "~var1 + var2 + var1:var2"){
  design_full = model.matrix(~var1 + var2 + var1:var2, data=DRIMSeq::sample(d))
}
colnames(design_full)
```

Run DRIMSeq

```{r run_DRIM_precision}
d = dmPrecision(d, design=design_full)
```
```{r run_DRIM_fit}
d = dmFit(d, design=design_full)
```
```{r run_DRIM_test}
d = dmTest(d, coef=colnames(design_full)[2])
```

Results

```{r result_DRIM}
res = DRIMSeq::results(d)
head(res)
```

Now at the transcript level

```{r tx_level_DRIM}
res.txp = DRIMSeq::results(d, level="feature")
head(res.txp)
```

We want to filter this by p-value
but we also need to make sure that any p-values which return an na value are set to 1 and not 0

```{r filter_drim_pval}
no.na = function(x) ifelse(is.na(x), 1, x)
res$pvalue = no.na(res$pvalue)
res.txp$pvalue = no.na(res.txp$pvalue)
```

We can plot the top (most confident, i.e. lowest p-value) switch

```{r plot_drim_pval}
idx = which(res$adj_pvalue < 0.05)[1]
res[idx,]
plotProportions(d, res$gene_id[idx], "var1")
```

### Final processing of DRIMSeq result with stageR

```{r stageR_drim}
pScreen = res$pvalue
strp = function(x) substr(x,1,15)
names(pScreen) = strp(res$gene_id)

pConfirmation = matrix(res.txp$pvalue, ncol=1)
rownames(pConfirmation) = strp(res.txp$feature_id)

tx2gene = res.txp[,c("feature_id", "gene_id")]
for (i in 1:2) tx2gene[,i] = strp(tx2gene[,i])

stageRObj = stageRTx(pScreen = pScreen,
                     pConfirmation = pConfirmation,
                     pScreenAdjusted = FALSE,
                     tx2gene = tx2gene)

stageRObj = stageWiseAdjustment(stageRObj, method="dtu", alpha=0.05)

suppressWarnings({
  drim.padj = getAdjustedPValues(stageRObj, order=FALSE, onlySignificantGenes = TRUE)
})

head(drim.padj)
```


### DEXSeq 

We can use the data we previously formatted as a starting point

N.b. even though the model uses "exon" here we are feeding it transcripts so it will still work

```{r DEXseqDataSet_from_DRIM}
sample.data = DRIMSeq::samples(d)
count.data = round(as.matrix(counts(d)[,-c(1:2)]))
dxd = DEXSeqDataSet(countData = count.data,
                    sampleData = sample.data,
                    design = ~sample + exon + var1:exon,
                    featureID = counts(d)$feature_id,
                    groupID = counts(d)$gene_id)
```

Now run the DEXSeq model

```{r run_DEX_estimateSizeFactors}
dxd = estimateSizeFactors(dxd)
```
```{r run_DEX_estimateDispersion}
dxd = estimateDispersions(dxd, quiet=TRUE)
```
```{r run_DEX_testForDTU}
dxd = testForDEU(dxd, reducedModel = ~sample + exon)
```

And obtain DEXSeqResult

```{r result_DEX}
dxr = DEXSeqResults(dxd, independentFiltering = FALSE)
qval = perGeneQValue(dxr)
dxr.g = data.frame(gene=names(qval),qval)
```

And as before we will screen with stageR

```{r stageR_DEX}
pConfirmation = matrix(dxr$pvalue, ncol=1)
dimnames(pConfirmation) = list(strp(dxr$featureID), "transcript")
pScreen = qval
names(pScreen) = strp(names(pScreen))
tx2gene = as.data.frame(dxr[,c("featureID", "groupID")])
for (i in 1:2) tx2gene[,i] = strp(tx2gene[,i])
```

```{r result_stageR_DEX}
stageRObj = stageRTx(pScreen=pScreen, 
                     pConfirmation = pConfirmation,
                     pScreenAdjusted = TRUE,
                     tx2gene=tx2gene)
stageRObj = stageWiseAdjustment(stageRObj, method="dtu", alpha = 0.05)

suppressWarnings({
  dex.padj = getAdjustedPValues(stageRObj, order=FALSE, onlySignificantGenes = TRUE)
})
```


And write the files out 

DRIMSeq files (initial (res and res.txp) and stageR-screened (drim.padj))

```{r writeout_DRIM}
#DRIMSeq outputs
#Initial result (Gene level)
write.csv(res, "drim_dtu_genes.csv")
#Tx level
write.csv(res.txp, "drim_dtu_transcripts.csv")

#stageR-screened
write.csv(drim.padj, "drim_dtu_stageR-screened.csv")
```

DEXSeq files (initial (dxr), genes (dxr.g), and stageR-screened (dex.padj))

```{r writeout_DEX}
#DEXSeq outputs
#Initial result
write.csv(dxr, "dex_dtu_all.csv")
#Gene list
write.csv(dxr.g, "dex_dtu_genes.csv")

#stageR-screened
write.csv(dex.padj, "dex_dtu_stageR-screened.csv")
```

Finally, save workspace image so I can come back to this if needed. 

```{r save_DTU_image}
save.image("DRIMDEX_workflow.RData")
```